### ISSUE TITLE
Allow the user to select which folders to include/exclude during full account refresh

### ISSUE THREAD

**Issue by leandro-lucarella-sociomantic** [owncloud/android#960]
Coming from #953, I would like to propose a new feature to let the user select which folders in the server they want to synchronize (as see in the Android client at all, I'm not talking about 2-way sync or anything like that).

For example, I have a lot of files in my server, but I only keep a small folder with a few files that I want to access in my phone. I don't want to be even able to access the other folders in my phone, only that one, so I would like to be able to set up the android client to only see that particular folder.

BTW, this is how the desktop client (at least in linux) works.

**Comment by derekblankmccoy** [comment]
This might work for you: Create a new user called mobile (or whatever you want). Share the folder with that user. Sign into the android client as that mobile user.

**Comment by masensio** [comment]
Hi @leandro-lucarella-sociomantic,

I'm confused with the description of this issue, because in #953 you said that you need to avoid to synchronize some folders, not to select which folders you want to sync.
 
(Comment in #953) 
leandro-lucarella-sociomantic commented 8 days ago
>No, I mean the exact opposite, to be able to tell the android client to absolutely ignore some folders. In the desktop client is the other way around, instead of having automatic access to the root directory, you set up which folders you want to sync to. So, I have like 3 top-level folders. 1 that is he only one is important to me to be accessible in the phone and with just a few files, and 2 that have a big number of subfolders. If I could tell the android client to only look at that small subfolder, that would let me use the owncloud client in android, which now for me is absolutely not an option.

If you need "Allow the user to select which folders to synchronize", this issue is duplicated (it is included in #3)

**Comment by leandro-lucarella-sociomantic** [comment]
> I'm confused with the description of this issue, because in #953 you said that you need to avoid to synchronize some folders, not to select which folders you want to sync.

Both white lists and black lists would be ideal ;-)

> If you need "Allow the user to select which folders to synchronize", this issue is duplicated (it is included in #3)

I think the confusion here comes from the terminology, I'm sorry if I'm using it incorrectly. By "synchronize", I don't mean "two-way sync" as #3 does. I'm just referring to folders being visible by the android client at all. I'm am talking about the current status of the android app, not about two-way sync.

All I want to do is to have a way to tell the android app to completely ignore some folders (black list) or to only show some folders, ignoring anything else (white list).

Does this clarify what I mean?

**Comment by derekblankmccoy** [comment]
@leandro-lucarella-sociomantic 
Just make a user and share the folder with that user, like I mentioned earlier, simple.

**Comment by leandro-lucarella-sociomantic** [comment]
@derekblankmccoy Thank's for the advice, but that doesn't work if you don't control the server and you can't create more users for yourself.

**Comment by derekblankmccoy** [comment]
Oh, the post says its your server

**Comment by leandro-lucarella-sociomantic** [comment]
> Oh, the post says its your server

Yes, I have my own server, but still, I'm addressing this issue in a more general way than just my particular personal problem. :)

**Comment by StevenChorkley** [comment]
+1
I want to sync 1Password with my phone.

**Comment by davidchisnall** [comment]
+1  Before I go on a trip I create a new folder with all of the information I'm going to need and I'd love to be able to have a one-tap way of ensuring that everything in that folder is on my phone / tablet.  I currently have to go through and individually select each item, select 'details' from the menu, and then tick 'keep up to date'.  When I get back, I need to do the same thing again in reverse.  Even having a 'keep up to date' checkbox on the list would reduce the number of taps to a third...

**Comment by gerroon** [comment]
Why cant we have keep up to date with folders as well? Having it only with files is very tedious unless  you are sharing only couple files.

**Comment by rperezb** [comment]
@davidchisnall on the Android app there is an option "download folder", if you do long press over a folder it is shown, if you select it, the behavior is what you are asking for: "have a one-tap way of ensuring that everything in that folder is on my phone / tablet"
The reverse, "keep-up-to-date" folders is still not implemented and it will come in future versions, as it is on the files.

**Comment by davivel** [comment]
@leandro-lucarella-sociomantic , I'm not sure if I understand your real necessity. Maybe you have enabled the automatic synchronization of your account in the system settings, and you don't want that the list of files is fully updated? Or do you have lots of files and foders in your root folder? Otherwise, I can't see how this feature could be useful.

**Comment by leandro-lucarella-sociomantic** [comment]
> Maybe you have enabled the automatic synchronization of your account in the system settings

Now it is disabled because of this issue, but yeah, I want to enable automatic sync.

> and you don't want that the list of files is fully updated? Or do you have lots of files and foders in your root folder

Yes, not in the root folder, but yes, I have a lot of files. As it says in #953, ~67K files and ~4K folders (which is the real problem, as the number of requests done against the server is `O(folders)`).

**Comment by davivel** [comment]
OK, I understand best now. Thanks.

**Comment by davidchisnall** [comment]
@rperezb that does seem to download it, but it's only half of the solution as it doesn't keep the files up to date.  In particular, there's no 'stop syncing all of the files in this folder' option.

I'm not sure why this should increase CPU load.  Protocols for low-bandwidth file change notifications for monitoring lots of files were a solved problem in the '90s.  There's no reason why the number of requests should be proportional to the number of files, especially when you have a directory structure for trivially aggregating change notifications.

**Comment by llucax** [comment]
Add to this issue the ability to choose in which local folder you want to sync every remote folder you select, and you also get #10, #209 and #780 fixed, and in a more general way.

BTW, this is how the (linux) desktop application works.

**Comment by gerroon** [comment]
I am confused. Is the "sync" option in the account setting supposed to sync all the downloaded files automatically or is it ssupposed to sync only the files that are tagged to be "up to date"? 

In any case auto sync does not seem to get new files from server automatically.

**Comment by davivel** [comment]
@gerroon, what do you mean exactly with ' the "sync" option'?

Thanls.

**Comment by gerroon** [comment]
@davivel 

We have the sync tick box in the account settings. I am wondering what it does exactly? I think that is where all the confusion comes from. 

As far as I know there is no "sync" as people understand it, like syncing folders etc . There is only keeping a particular file up to date if the user tagged that file in the OC client.

**Comment by davivel** [comment]
Enabling that option means that, whenever the Android system feels interesting, a full refresh of the account is triggered. When it happens exactly is quite unknown, since there is no detailed documentation about it (indeed, we should check this again, maybe it improved from the last time).

The full refresh means:
  - The app tries to fetch from the server the metadata of all the files & folders in the account. All the tree. Metadata: name, size, MIME-type, modification date. NOT the contents.
  - The app tries to synchronize the CONTENTS of all the files that are marked to be kept-in-sync.
  - If there is an error, the process does not go deeper in that branch of the files tree, and tries to go on through a sibling branch.
  - If there is another error, the process stops with an error.

Same process occurs when the action "Refresh full account" is started from the menu in the top bar.

I'll search for some fixed location to copy & paste this explanation.

**Comment by gerroon** [comment]
@davivel .


Thanks for the detailed explanation

**Comment by maystar** [comment]
I can confirm that the automatic sync for the whole account is not usable for owncloud folders with many files. I store my whole music and video library in external folders inside my account (of course they are not synchronized with the desktop clients). I couldn't enable the sync feature for the owncloud account in no prior app version. The sync process take hours, drains the battery and do not successfully finish. I even suspect android to queue other sync accounts until the active one is finished. 
A feature for ignoring some directories would be great! It could look similar to the folder selection of the recent desktop client.

**Comment by maystar** [comment]
Btw. perhaps the name of this issue should be more like "Allow the user to select which folders to include/exclude during full account refresh".

Of course it would be great if the refresh is fast enough to handle also huge directories. But I think this is more difficult to implement than a white/blacklist.

**Comment by llucax** [comment]
> I store my whole music and video library in external folders inside my account (of course they are not synchronized with the desktop clients).

That's exactly my use case too, add on top of that all the photos I've taken in my life :)

**Comment by llucax** [comment]
> Btw. perhaps the name of this issue should be more like "Allow the user to select which folders to include/exclude during full account refresh".

Yes, thank you, I think that might reduce the amount of confusion around this issue. I can't update the title of the issue though, I guess a contributor has to do it.

**Comment by maystar** [comment]
If I trigger the account refresh manually it will finish after 1,5 h with this messages:
```D/SyncManager(  722): failed sync operation user@server u0 (owncloud), org.owncloud, SERVER, latestRunTime 2129857, reason: AutoSync, SyncResult: syncAlreadyInProgress: true stats []
D/SyncManager(  722): failed sync operation user@server u0 (owncloud), org.owncloud, LOCAL, latestRunTime 2199253, reason: 10114, SyncResult: syncAlreadyInProgress: true stats []```
At this point not all directories and files were refreshed.

**Comment by malkomich** [comment]
You can track this issue in the following discussion: https://github.com/owncloud/android/issues/1136, which try to handle all features related to folder synchronization.

**Comment by leandro-lucarella-sociomantic** [comment]
So you'll basically be able to map a folder in the phone to a folder in the owncloud server after that issue is fixed? That's the idea?

**Comment by davivel** [comment]
No, what you describe is #1336 + #1663 (for contents under the local ownCloud folder) , and #744, for contents out of local ownCloud folder.

**Comment by leandro-lucarella-sociomantic** [comment]
OK, so with #1136 the sync function of owncloud will only refresh the selected folders, not everything that's available in the owncloud server. Is that it? I'm sorry but I'm having trouble to see the relationship between this issue and #1136 (or how #1136 will address this issue).

**Comment by llucax** [comment]
@davivel ping?

**Comment by davivel** [comment]
Sorry, @llucax .

I guess your problem is you have tons of files and you don't want to waste battery & network synchronizing them, but you still want to use automatic synchronization of part of the file tree, right?

You are right, it's a different detail. We'll keep it as a separate case.

**Comment by leandro-lucarella-sociomantic** [comment]
Yup, some files I only care to be able to get on-demand, and some files I want to keep in sync. Even completely ignoring some files would be fine (as I do with the desktop client, where I can just pick which files to synchronize).

**Comment by DeepDiver1975** [comment]
🤖 Automated issue triage by Claude Code review agent.

This issue was opened before 2020 and has had no activity since then, so we are
closing it as outdated. If this is still relevant on a current ownCloud release,
please **reopen** (or open a fresh issue) with up-to-date details — thank you\!
